core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -1917,7 +1918,9 @@ private RexNode simplifyCast(RexCall e) {
       if (RexUtil.isLosslessCast(intExpr.getType(), operand.getType())
           && (e.getType().getSqlTypeName() == operand.getType().getSqlTypeName()
           || e.getType().getSqlTypeName() == SqlTypeName.CHAR
-          || operand.getType().getSqlTypeName() != SqlTypeName.CHAR)) {
+          || operand.getType().getSqlTypeName() != SqlTypeName.CHAR)
+          && SqlTypeAssignmentRule.instance()
+          .canApplyFrom(intExpr.getType().getSqlTypeName(), e.getType().getSqlTypeName())) {
         return rexBuilder.makeCast(e.getType(), intExpr);
```

Comment:
Should be `SqlTypeCoercionRule` ?

Reply:
It seems you're right. BYW, what's the difference between `SqlTypeCoercionRule ` and `SqlTypeAssignmentRule `?

Reply:
See the java doc which already gave good explanation.